integrations/datadog/client.py

```diff
@@ -233,6 +297,176 @@ async def get_slo_history(
+            tasks = []
+            for env_to_fetch in envs_to_fetch:
+                params = {"env": env_to_fetch, "service": service_id}
+                query_with_values = self._create_query_with_values(
+                    f"{query}{{service:{service_id},env:{env_to_fetch}}}", params
+                )
+
+                end_time = int(time.time())
+                start_time = end_time - (time_window * 60)
+
+                url = f"{self.api_url}/api/v1/query?from={start_time}&to={end_time}&query={query_with_values}"
+
+                # Create a task for each fetch operation
+                task = asyncio.create_task(
+                    self._fetch_with_rate_limit_handling(
+                        url, semaphore=self._metrics_semaphore
+                    )
+                )
+                tasks.append(task)
+
+            # Gather results concurrently
+            results = await asyncio.gather(*tasks)
+
+            # Process and yield results
+            metrics = []
+            for result, env_to_fetch in zip(results, envs_to_fetch):
+                # Update result with metadata
+                result.update(
+                    {
+                        "__service": service_id,
+                        "__metric": self.extract_metric_name(query),
+                        "__query_id": f"{query}/service:{service_id}/env:{env_to_fetch}",
+                        "__query": query,
+                        "__env": env_to_fetch,
+                    }
+                )
+                metrics.append(result)
+
+            yield metrics
+
+    async def get_metrics(
+        self,
+        query: str,
+        env: str = "*",
+        service: str = "*",
+        time_window: int = FETCH_WINDOW_TIME_IN_MINUTES,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        """
+        Fetches metrics for specified services and environment.
+
+        Args:
+            query (str): The Datadog query string to execute.
+            env (str): The environment to filter by (default: "prod").
+            service (str): The service ID to filter by, or "*" to fetch metrics for all services.
+            time_window (int): Time window in minutes for fetching metrics (default: FETCH_WINDOW_TIME_IN_MINUTES).
+
+        Yields:
+            AsyncGenerator[list[dict[str, Any]], None]: Each individual metric as it's fetched.
+        """
+
+        envs_to_fetch = (
+            [env] if env != "*" else self.get_env_tags(await self.get_tags())
+        )
+        if not envs_to_fetch:
+            logger.warning("No environments found, exiting...")
+            return
+
+        if service == "*":
+            async for service_list in self.get_services():
+                async for metric in self._fetch_metrics_for_services(
+                    query, envs_to_fetch, service_list, time_window
+                ):
+                    yield metric
+        else:
+            result = await self.get_single_service(service)
```

Comment:
just to make sure, it returns empty result? and not an error of 404?

Reply:
nice catch. I have revised this